### PR TITLE
Angular2 override

### PR DESCRIPTION
**Pull request by guybedford**
Override for https://github.com/jspm/registry/issues/795.

//cc @amcdnl

It would be good to get confirmation from others using Angular if this is all good. @robwormald perhaps you could take a look here or suggest someone who could?

### CODE REVIEW

**Comment by amcdnl** [comment]
If it helps, you can see the overrides in action w/ this repo - https://github.com/amcdnl/angular2-upgrading

**Comment by amcdnl** [comment]
Can we go ahead and merge? I know @robwormald is very busy and this would really help me out a lot locally.

**Comment by guybedford** [comment]
This is a change that will need to remain backwards compatible, and should be considered part of the Angular API - I'd only want to merge with confirmation from someone on the Angular team.

**Comment by amcdnl** [comment]
Alright I understand, anyone else we can hit up to review?

**Comment by guybedford** [comment]
@jeffwhelpley this is quite an important registry entry to get Angular2 correctly configured for jspm. Do you know anyone from Angular who may be able to review this PR before we go ahead here?

**Comment by jeffwhelpley** [comment]
@guybedford yes, the best ones to look at this would be @gdi2290 or @jeffbcross. I will ping them on slack as well to make sure they see this and check it out.

**Comment by guybedford** [comment]
That would be amazing, thanks.

File: package-overrides/npm/angular2@2.0.0-beta.8.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "dependencies": {
+    "zone.js": "npm:zone.js@0.6.2",
```
Comment (jeffbcross) [inline package-overrides/npm/angular2@2.0.0-beta.8.json]: I believe beta.8 depends on zone.js@0.5.15

**Comment by gdi2290** [comment]
The current version if Angular 2 beta 8 is using zone.js 0.5.15

File: package-overrides/npm/angular2@2.0.0-beta.8.json (hunk 1 above)
Comment (jeffbcross) [inline package-overrides/npm/angular2@2.0.0-beta.8.json]: And why are es6-promise and es6-shim not included? See package.json: https://github.com/angular/angular/blob/master/modules/angular2/package.json

Comment (amcdnl) [inline package-overrides/npm/angular2@2.0.0-beta.8.json]: @jeffbcross are those really required? I'm running babel which has these polyfilled already.

Comment (amcdnl) [inline package-overrides/npm/angular2@2.0.0-beta.8.json]: Its not a big deal to just leave these out altogether, all we really need is:
https://github.com/jspm/registry/blob/angular/package-overrides/npm/angular2%402.0.0-beta.8.json#L13 and the `empty` flag

**Comment by robwormald** [comment]
sorry, didn't see this earlier. I believe JSPM should remain consistent with the NPM peer deps, so really the override should be similar to https://github.com/angular/angular/blob/master/modules/angular2/package.json#L12-L16

I'm happy to discuss eventually trimming that, but let's stick to what we know works, for the moment - otherwise i'll end up with 5000 tweets about it when it inevitably breaks... Most users are using TS, so we can't rely on the babel coverage. Babel uses can certainly choose not to import them if they don't need them, but I don't think there's any harm in the extra install there (as opposed to the original dozen+ extraneous deps, which is why the original issue got opened up on our end)

**Comment by amcdnl** [comment]
@robwormald thats totally fine. The big thing we REALLY need is:

```
"map":{
    "crypto": "@empty"
  },
  "jspmNodeConversion": false
```

@guybedford : lets ditch the rest and just do the above if thats ok?

**Comment by robwormald** [comment]
where is crypto coming from? reflect-metadata? maybe that's where the map/metaconfig should be done then?

**Comment by amcdnl** [comment]
@robwormald cyrpto comes from angular when its running in Node mode.

**Comment by guybedford** [comment]
Thanks for all the feedback here, much appreciated for the help to make sure this goes smoothly, as these overrides do form part of the "breakable" API.

* I was able to remove the `crypto` mapping since we have `jspmNodeConversion: false` set jspm will not attempt to automatically install crypto. The only issue will be that modules that require crypto will fail. Looking through the codebase this was in bundles/angular2-polyfills.js and ts/typings/node/node.d.ts. @amcdnl I take it you were loading the angular2-polyfills.js file then that you needed this, which was then giving a loading error? In that case, it may be better to set a `meta: { "bundles/angular2-polyfills.js": { cjsRequireDetection: false } }` to disable static analysis, but still allow crypto to resolve if it does exist in the registry. But that feature won't work in jspm 0.16 unfortunately. Otherwise I can add back the empty map too.
* I've set the dependencies as both dependencies and peerDependencies, so that in jspm 0.16 the dependencies will install as normal dependencies, and in jspm 0.17 the dependencies will install as peerDependencies.
* I've also set the dependencies to semver ranges. While this specific beta release is locked, the next beta will bump the ranges, and this override will continue to apply. That could potentially cause future releases to break as this override wouldn't update with them. Because jspm is locked by default, it shouldn't be as much of a problem to allow semver ranges in the package.json here.

**Comment by amcdnl** [comment]
@guybedford -

> The only issue will be that modules that require crypto will fail. Looking through the codebase this was in bundles/angular2-polyfills.js and ts/typings/node/node.d.ts. @amcdnl I take it you were loading the angular2-polyfills.js file then that you needed this, which was then giving a loading error

Yes, polyfills are required to run Angular2 so when I include them it brings in crypto.

> In that case, it may be better to set a meta: { "bundles/angular2-polyfills.js": { cjsRequireDetection: false } } to disable static analysis, but still allow crypto to resolve if it does exist in the registry. But that feature won't work in jspm 0.16 unfortunately. Otherwise I can add back the empty map too.

I'm not running 17 yet, I'm waiting for it to RC before hopping on. We have our bundler builders and whatnot that will need to be upgraded and wanted things to stabilize out before I move.

Here is a small demo of A2 running with JSPM = https://github.com/amcdnl/angular2-upgrading

**Comment by robwormald** [comment]
@amcdnl generally, you shouldn't be importing the polyfills file anyway - zone needs to do its thing before the module loader boots, or you can get caught in all kinds of tricky zone issues.

**Comment by guybedford** [comment]
@amcdnl let me know if you can work with Rob's suggestion here?

**Comment by amcdnl** [comment]
@robwormald hmmm ... that might explain some of my change detection issues!
@guybedford Yes, I'm good. I mainly only wanted the `jspmNodeConversion:false` and `crypto:empty`.

Lets call it done! :)

**Comment by guybedford** [comment]
Please let's not call it done without 100% certainty. This is like publishing any release - the API should be carefully considered to be sure we can maintain backwards compatibility.

Specifically what still worries me is if the angular2-polyfills.js file shouldn't be loaded manually, what will that affect here specifically? What usually loads this file?

**Comment by amcdnl** [comment]
@guybedford no problem! So @robwormald correct me if I'm wrong but if you don't load it in the right order the change detection won't trigger correctly. Angular2 hooks into events of the browser and without loaded in the correct order, angular core can't update correctly.

**Comment by guybedford** [comment]
I've gone through this again against a test require of `System.import('angular2/core.js')` and `System.import('angular2/platform/browser.js')` in both jspm 0.16 and jspm 0.17 in both the browser and Node.

I was able to make the further simplifications:
* Add an override for zone.js to use the single-file build over the multifile unlinked version, this is for faster loading
* Update the reflect-metadata override to pull in fewer dependencies as it doesn't need Node shims
* In the jspm 0.17 case, automatically load the `reflect-metadata` shim when loading `src/core/util/decorators.js` in Angular. I really don't understand why this isn't already pulled in automatically?

My experience using Angular is very limited, so it would help to get feedback on further edge cases here... what other requires should be checked here beyond the above? Is `angular-polyfills.js` loaded manually by users, or is it not necessary to load at all?

There's likely a lot more that could be done here to improve this loading experience. Things like checking code branching between Node and browser (there doesn't seem to be any branching at the module level), using builds of private modules like I've done with the zone.js override to avoid many separate files (rxjs seems suitable for the separate files though as it is so heavily modular, but maybe more could be done here in Angular itself?). Always happy to discuss this stuff further.

Another thing worth considering is whether we should use the TypeScript source for angular by default, and then auto-configure the TypeScript plugin to apply correctly with Type checking in the Angular config. That might be quite a nice workflow too, depending on performance.

As soon as this is merged, I'll also post PRs to all the angular repos to have this config in the original package.json files too, although no pressure for merging those.

**Comment by amcdnl** [comment]
@guybedford - awesome notes! 

Note: We should probably bump this to beta9 now since thats the latest. There are no major changes between versions, bug fixes and router additions.

**Comment by guybedford** [comment]
Ok, I'm merging this. If there are any angry users, just let me know :)

**Comment by guybedford** [comment]
~~Angry~~ user number 1 - https://github.com/jspm/jspm-cli/issues/1651.

**Comment by guybedford** [comment]
I've created a correction PR at https://github.com/jspm/registry/pull/846.

**Comment by amcdnl** [comment]
:( - we've been running and not had this issue.

**Pull request merged**